django/contrib/admin/options.py

```diff
@@ -2055,12 +2055,6 @@ def get_formset(self, request, obj=None, **kwargs):
         can_add = self.has_add_permission(request, obj) if request else True
 
         class DeleteProtectedModelForm(base_model_form):
-            def __init__(self, *args, **kwargs):
-                super(DeleteProtectedModelForm, self).__init__(*args, **kwargs)
-                if not can_change and not self.instance._state.adding:
-                    self.fields = {}
-                if not can_add and self.instance._state.adding:
-                    self.fields = {}
```

Comment:
Are we sure that including fields from now on won't allow a malicious user to provide this data from now on?

Maybe there's tests for these cases but I felt like removing all fields in `__init__` was _safer_ than relying on side effects through `has_changed`?

Reply:
As far as I know, formsets always ignore forms that haven't changed.

The test that fails with these lines removed is:
https://github.com/django/django/blob/64e1a271f50d921a54388539b6ff7102a31c3d29/tests/admin_views/tests.py#L3858-L3902

Reply:
Thanks for looking that up @timgraham, as long as have tests asserting that this is working as expected it should be fine.